hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/oauth2/WorkloadIdentityTokenProvider.java

```diff
@@ -0,0 +1,144 @@
+    Preconditions.checkNotNull(tokenFile, "tokenFile");
+
+    this.authEndpoint = authority + tenantId + OAUTH2_TOKEN_PATH;
+    this.clientId = clientId;
+    this.tokenFile = tokenFile;
+  }
+
+  @Override
+  protected AzureADToken refreshToken() throws IOException {
+    LOG.debug("AADToken: refreshing token from JWT Assertion");
+    String clientAssertion = getClientAssertion();
+    AzureADToken token = getTokenUsingJWTAssertion(clientAssertion);
+    tokenFetchTime = System.currentTimeMillis();
+    return token;
+  }
+
+  /**
+   * Gets the Azure AD token from a client assertion in JWT format.
+   * This method exists to make unit testing possible.
+   *
+   * @param clientAssertion the client assertion.
+   * @return the Azure AD token.
+   * @throws IOException if there is a failure in connecting to Azure AD.
+   */
+  @VisibleForTesting
+  AzureADToken getTokenUsingJWTAssertion(String clientAssertion) throws IOException {
+    return AzureADAuthenticator
+        .getTokenUsingJWTAssertion(authEndpoint, clientId, clientAssertion);
+  }
+
+  /**
+   * Checks if the token is about to expire as per base expiry logic.
+   * Otherwise, try to expire if enough time has elapsed since the last refresh.
+   *
+   * @return true if the token is expiring in next 1 hour or if a token has
+   * never been fetched
+   */
+  @Override
+  protected boolean isTokenAboutToExpire() {
+    return super.isTokenAboutToExpire() || hasEnoughTimeElapsedSinceLastRefresh();
+  }
+
+  /**
+   * Checks to see if enough time has elapsed since the last token refresh.
+   *
+   * @return true if the token was last refreshed more than an hour ago.
+   */
+  protected boolean hasEnoughTimeElapsedSinceLastRefresh() {
+    if (getTokenFetchTime() == -1) {
+      return true;
+    }
+    boolean expiring = false;
+    long elapsedTimeSinceLastTokenRefreshInMillis =
+        System.currentTimeMillis() - getTokenFetchTime();
+    // In case token is not refreshed for 1 hr or any clock skew issues,
+    // refresh token.
+    expiring = elapsedTimeSinceLastTokenRefreshInMillis >= ONE_HOUR
+        || elapsedTimeSinceLastTokenRefreshInMillis < 0;
+    if (expiring) {
+      LOG.debug("JWTToken: token renewing. Time elapsed since last token fetch:"
+          + " {} milliseconds", elapsedTimeSinceLastTokenRefreshInMillis);
+    }
+    return expiring;
+  }
+
+  /**
+   * Gets the client assertion from the token file.  The token in the file
+   * is automatically refreshed by Azure at least once every 24 hours.
+   * See <a href="https://azure.github.io/azure-workload-identity/docs/faq.html#does-workload-identity-work-in-disconnected-environments">
+   * Azure Workload Identity FAQ</a>.
+   *
+   * @return the client assertion.
+   * @throws IOException if the token file is empty.
+   */
+  private String getClientAssertion()
+      throws IOException {
+    File file = new File(tokenFile);
+    String clientAssertion = FileUtils.readFileToString(file, "UTF-8");
```

Comment:
Why not do a try catch and throw the actual exception that occurred during the file read? 
and if no value is present throw below IOE.

Reply:
Taken